### ISSUE TITLE
model.fit can crash!

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#217]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/217) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

```python
import mass
import numpy as np
import lmfit

counts = np.array([2, 4, 2, 4], dtype=np.int64)  
bin_centers = np.array([8009.07622011, 8009.57622011,
       8010.07622011, 8010.57622011])

model = mass.spectra["CuKAlpha"].model()
params = model.guess(bin_centers=bin_centers, data=counts)
params["dph_de"].set(1, min=0.1, max=10, vary=False)
params["fwhm"].set(4)
params["peak_ph"].set(8009.57622011)
params["integral"].set(0)
params["background"].set(2.0000000000000004)
params["bg_slope"].set(0)
r = model._residual(params=params, bin_centers=bin_centers, data=counts, weights=None)
assert not any(np.isnan(r))
```

This is a minimum example of how `model._residual` can return nan values, which makes `model.fit` crash. I suspect the issue is one of floating point rounding errors allowing negative numbers where perfect math would make them impossible.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Here is an example of how the math in the failing function works out…. the next step would be `sqrt(r22)` which is now negative, but should have been zero.

```python
y,data=np.array([2.0000000000000004]),np.array([2])
r2=y-data
tmp=np.log(data/y)
r22=r2+data*tmp

print(f"{r2=} + {data*tmp=} = {r22=}")

# gives r2=array([4.4408921e-16]) + data*tmp=array([-4.4408921e-16]) = r22=array([-9.86076132e-32])
```

‌

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

The actual failure occurs here

[https://bitbucket.org/joe\_fowler/mass/src/4205edf2981db59aed3218f514facfca53be7d31/mass/calibration/line\_models.py?at=master#lines-21](https://bitbucket.org/joe_fowler/mass/src/4205edf2981db59aed3218f514facfca53be7d31/mass/calibration/line_models.py?at=master#lines-21)

presumably we need to check for very small negative numbers before the sqrt? or the nonzero check could be for numbers larger than some threshold like `1e-15`?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I’m working this problem with you. As I said in our chat, the first example does not fail on my Mac, and the example in your first comment does not print that r22 is a negative number. It prints the desired result: `r22=array([0.])`

Proposal: I’ll add your first example as a unit test and see if it fails on the Bitbucket pipeline \(even though it doesn’t fail on my Mac\). If that happens, then we can be sure that our fix is a real fix.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Don't take sqrt of negative values

Fixes #217. Fix some linter warnings (unused variables).

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
